File: modules/mynlp-classification/src/main/java/com/mayabot/nlp/classification/FasttextClassification.java
package com.mayabot.nlp.classification;

import com.google.common.base.CharMatcher;
import com.google.common.base.Splitter;
import com.google.common.collect.Lists;
import com.mayabot.mynlp.fasttext.*;
import com.mayabot.nlp.segment.LexerReader;
import com.mayabot.nlp.segment.Lexers;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.List;

/**
 * 使用Fasttext算法的文本分类工具
 * <p>
 * 文本文件格式
 *
 * @author jimichan
 */
public class FasttextClassification {

    /**
     * 训练Fasttext模型
     */
    public static FastText train(File file, int dim, double learnRate, int epoch) throws Exception {
        TrainArgs trainArgs = new TrainArgs();
        trainArgs.setDim(dim);
        trainArgs.setLr(learnRate);
        trainArgs.setEpoch(epoch);

        return train(file, trainArgs);
    }

    /**
     * 训练Fasttext模型
     */
    public static FastText train(File file, TrainArgs trainArgs) throws Exception {

        Splitter splitter = Splitter.on(CharMatcher.whitespace()).omitEmptyStrings().trimResults();


        LexerReader analyzer = Lexers.coreBuilder().build().filterReader(true, true);

        FileTrainExampleSource source = new FileTrainExampleSource(
                new WordSplitter() {
                    @NotNull
                    @Override
                    public List<String> split(String text) {
                        List<String> result = Lists.newArrayList();
                        splitter.split(text).forEach(part -> {
                            if (part.startsWith("__label__")) {
                                result.add(part);
                            } else {
                                for (String word : analyzer.scan(part).toWordSequence()) {
                                    result.add(word);
                                }
                            }
                        });
                        return result;
                    }
                }
                ,
                file
        );
        return FastText.train(source, ModelName.sup, trainArgs);
    }


    /**
     * 使用乘积量化压缩Fasttext模型，稍微损失一些精度.
     *
     * @param fastText
     * @return FastText
     * @throws Exception
     */
    public static FastText compress(FastText fastText) throws Exception {
        return FastText.quantize(fastText);
    }

    /**
     * 便捷方法，返回预测的第一个结果
     *
     * @param model
     * @param tokenizer 分词器
     * @param text      需要分析的原始文本
     * @return 分类目标. null表示没有结果
     */
    public static List<FloatStringPair> predict(FastText model, LexerReader tokenizer, String text, int top) {
        List<String> inputList = Lists.newArrayList(tokenizer.scan(text).toWordSequence());
        if (inputList.isEmpty()) {
            return null;
        }

        return model.predict(inputList, top);
    }

    /**
     * 预测一个唯一的目标，要求概率大于0.5f
     *
     * @param model
     * @param tokenizer
     * @param text
     * @return FloatStringPair
     */
    public static FloatStringPair predictOne(FastText model, LexerReader tokenizer, String text) {
        List<String> inputList = Lists.newArrayList(tokenizer.scan(text).toWordSequence());
        if (inputList.isEmpty()) {
            return null;
        }

        List<FloatStringPair> list = model.predict(inputList, 5);

        if (list.isEmpty()) {
            return null;
        }

        FloatStringPair first = list.get(0);

        if (first.first < 0.5f) {
            return null;
        } else {
            return first;
        }
    }

    private static LexerReader lexerReader;

    public static List<FloatStringPair> predict(FastText model, String text, int top) {

        if (lexerReader == null) {
            lexerReader = Lexers.coreBuilder().build().filterReader(true,true);
        }

        return predict(model, lexerReader, text,top);
    }

    public static FloatStringPair predictOne(FastText model, String text) {

        if (lexerReader == null) {
            lexerReader = Lexers.coreBuilder().build().filterReader(true, true);
        }

        return predictOne(model, lexerReader, text);
    }
}

File: modules/mynlp-classification/src/main/java/com/mayabot/nlp/classification/FasttextClassification.kt
package com.mayabot.nlp.classification

import com.mayabot.nlp.fasttext.FastText
import com.mayabot.nlp.segment.Lexer
import com.mayabot.nlp.segment.LexerReader
import com.mayabot.nlp.segment.Lexers
import java.io.File

class FasttextClassification(val lexer: LexerReader,val fastText: FastText) {

    /**
     * 文本分类
     */
    @JvmOverloads
    fun classification(text: String,k:Int = 5,threshold:Float=0.0f): List<Pair<String,Float>> {
        val words = lexer.scan(text).toWordSequence()
        val target = fastText.predict(words,k,threshold)
        return target.map { it.label to it.score }
    }

    companion object{

        /**
         * 处理没有分词的语料
         * __label__xxxx 语料文本,语料文本，语料文本
         */
        @JvmOverloads
        @JvmStatic
        fun prepareBySegment(from: File,
                             to: File,
                             label:String = "__label__",
                             lexer: LexerReader = Lexers.coreBuilder().build().filterReader(true, true)) {

            fun processLine(line:String): String{
                val list = ArrayList<String>()
                line.split(" ").forEach { part->
                    if(part.startsWith(label)){
                        list += part
                    }else{
                        lexer.scan(part).toWordSequence().forEach { word->
                            list += word
                        }
                    }
                }
                return list.joinToString(" ")
            }
            from.useLines { lines->
                to.bufferedWriter(Charsets.UTF_8).use { writer->
                    lines.forEach { line->
                        writer.write(processLine(line))
                        writer.write("\n")
                    }
                }
            }
        }
    }

}